src/openms/include/OpenMS/FILTERING/TRANSFORMERS/SpectraMerger.h

```diff
@@ -436,6 +676,236 @@ namespace OpenMS
+        std::vector<double> mz_positions_all;    // m/z positions from all spectra
+        for (std::vector<std::pair<Size, double> >::const_iterator it2 = it->second.begin(); it2 != it->second.end(); ++it2)
+        {
+          // loop over m/z positions
+          for (typename MapType::SpectrumType::ConstIterator it_mz = exp[it2->first].begin(); it_mz < exp[it2->first].end(); ++it_mz)
+          {
+            mz_positions_all.push_back(it_mz->getMZ());
+          }
+        }
+        
+        sort(mz_positions_all.begin(), mz_positions_all.end());
+        
+        std::vector<double> mz_positions;    // positions at which the averaged spectrum should be evaluated
+        std::vector<double> intensities;
+        double last_mz = -1000.0;    // last m/z position pushed through from mz_position to mz_position_2
+        double delta_mz(mz_binning_width);    // for m/z unit Da
+        for (std::vector<double>::iterator it_mz = mz_positions_all.begin(); it_mz < mz_positions_all.end(); ++it_mz)
+        {
+          if (mz_binning_unit == "ppm")
+          {
+            delta_mz = mz_binning_width * (*it_mz) / 1000000;
+          }
+          
+          if (((*it_mz) - last_mz) > delta_mz)
+          {
+            mz_positions.push_back(*it_mz);
+            intensities.push_back(0.0);
+            last_mz = *it_mz;
+          }
+        }
+        
+        // loop over spectra in blocks
+        for (std::vector<std::pair<Size, double> >::const_iterator it2 = it->second.begin(); it2 != it->second.end(); ++it2)
+        {
+          SplineSpectrum spline(exp[it2->first]);
+          SplineSpectrum::Navigator nav = spline.getNavigator();
+          
+          // loop over m/z positions
+          for (int i = 0; i < mz_positions.size(); ++i)
+          {
+            if ((spline.getMzMin() < mz_positions[i]) && (mz_positions[i] < spline.getMzMax()))
+            {
+              intensities[i] += nav.eval(mz_positions[i]) * (it2->second);    // spline-interpolated intensity * weight
+            }
+          }
+        }       
+        
+        // update spectrum
+        typename MapType::SpectrumType average_spec = exp[it->first];
+        average_spec.clear(false);    // Precursors are part of the meta data, which are not deleted.
+        //average_spec.setMSLevel(ms_level);
+        
+        // refill spectrum
+        for (int i = 0; i < mz_positions.size(); ++i)
+        {
+          typename MapType::PeakType peak;
+          peak.setMZ(mz_positions[i]);
+          peak.setIntensity(intensities[i]);
+          average_spec.push_back(peak);
+        } 
+        
+        // store spectrum temporarily
+        exp_tmp.addSpectrum(average_spec);
+      }
+      
+      logger.endProgress();
+      
+      // loop over blocks
+      int n(0);
+      //typename MapType::SpectrumType empty_spec;
+      for (AverageBlocks::ConstIterator it = spectra_to_average_over.begin(); it != spectra_to_average_over.end(); ++it)
```

Comment:
I only want to update spectra of one particular MS level. That's done spectrum by spectrum.

@hroest  Do you want the generation of exp_tmp changed, or the final exp_tmp -> exp transfer?